### PR TITLE
Fix diversity

### PR DESCRIPTION
**Pull request by AWBroscius**
Address post pull request task list on [Issue 130](https://github.com/theochem/DiverseSelector/issues/130)

### CODE REVIEW

File: DiverseSelector/diversity.py [hunk 1]
```diff
@@ -96,53 +100,46 @@ def entropy(x: np.ndarray) -> float:
...
-    y = np.empty(x.shape)
-    for i in range(0, len(x)):
-        for j in range(0, len(x[0])):
-            if x[i, j] != 0:
-                y[i, j] = 1
-            else:
-                y[i, j] = 0
     # initialize variables
     length = len(x[0])
     n = len(x)
```
Comment (FarnazH) [inline DiverseSelector/diversity.py:118]: These two variables seem to represent the same value. Any reason we need both?

File: DiverseSelector/diversity.py [hunk 2]
```diff
@@ -96,53 +100,46 @@ def entropy(x: np.ndarray) -> float:
...
-    # count bits in fingerprint
-    for i in range(0, length):
-        val.append(sum(y[:, i]))
-    ans = np.sort(val)
+    # count bits in fingerprint and order them
+    counts = np.count_nonzero(x, axis=0)
+    counts = np.sort(counts)
     # sum entropy calculation for each feature
+    if 0 in counts:
+        raise ValueError
```
Comment (FarnazH) [inline DiverseSelector/diversity.py:125]: There should be a message to clarify what went wrong.

**Review by FarnazH — CHANGES_REQUESTED** [review]

File: DiverseSelector/diversity.py (hunk 1 above)
Comment (AWBroscius) [reply DiverseSelector/diversity.py]: They are not the same value, n is the number of molecules and length is the number of features. I've renamed them in my latest commit to make this clearer.

Comment (FarnazH) [reply DiverseSelector/diversity.py]: Yup, you are right. they are the number of rows and columns. It is a strange way of getting rows and columns, but I will fix it myself.

File: DiverseSelector/diversity.py [hunk 3]
```diff
@@ -42,7 +42,8 @@
 
 def compute_diversity(
     features: np.array,
-    div_type: str = "total_diversity_volume",
+    div_type: str = "entropy",
+    library: np.array = None,
```
Comment (FanwangM) [inline DiverseSelector/diversity.py:46]: The wording of `library` is too generic and can be confusing as we often use it to denote software packages. Maybe change it to `features` or something that's clear?

File: DiverseSelector/diversity.py [hunk 4]
```diff
@@ -68,12 +69,13 @@ def compute_diversity(
...
         "shannon_entropy": shannon_entropy,
         "wdud": wdud,
-        "total_diversity_volume": total_diversity_volume,
         "gini_coefficient": gini_coefficient,
     }
 
     if div_type in func_dict:
         return func_dict[div_type](features)
+    elif div_type == "hypersphere_overlap_of_subset":
+        return hypersphere_overlap_of_subset(library, features)
```
Comment (FanwangM) [inline DiverseSelector/diversity.py:78]: The usage of `features` is a little bit confusing here. Maybe try `feature_subset`? Someone would wonder what are `library` and `features` here. Moreover, one may assume the features denotes the full dature matrix. Anything that makes it clear works.

**Review by FanwangM — CHANGES_REQUESTED** [review]

**Comment by FanwangM** [comment]
I am good with the current PR. 

We are missing one test for `explicit_diversity_index`. We can use `RDKit` to generate the small dataset in Figure 4A and use that as a test,  https://pubs.acs.org/doi/10.1021/ci060074f. This may be a good exercise for someone who wants to learn more about `RDKit`.

Do you have any suggestions and is there anything we are still missing for the PR? I will fix the conflicts once you think this is good to merge. @FarnazH

**Review by FanwangM — CHANGES_REQUESTED** [review]
I am good with most of the changes in this PR. Do you have any other comments? @FarnazH

**Comment by FanwangM** [comment]
The only problem is coding style, which I will address it later.

**Pull request merged**